File: src/shared/go/pkg/health/health.go
package health

import (
	"net/http"
	"sync/atomic"

	"github.com/gin-gonic/gin"
)

type Checker struct {
	ready   atomic.Bool
	started atomic.Bool
}

func New() *Checker {
	return &Checker{}
}

func (c *Checker) SetReady(ready bool) {
	c.ready.Store(ready)
}

func (c *Checker) SetStarted(started bool) {
	c.started.Store(started)
}

func (c *Checker) RegisterRoutes(r *gin.Engine) {
	h := r.Group("/health")
	h.GET("/ready", c.readyHandler)
	h.GET("/live", c.liveHandler)
	h.GET("/startup", c.startupHandler)
}

func (c *Checker) readyHandler(ctx *gin.Context) {
	if c.ready.Load() {
		ctx.JSON(http.StatusOK, gin.H{"status": "ready"})
		return
	}
	ctx.JSON(http.StatusServiceUnavailable, gin.H{"status": "not_ready"})
}

func (c *Checker) liveHandler(ctx *gin.Context) {
	ctx.JSON(http.StatusOK, gin.H{"status": "alive"})
}

func (c *Checker) startupHandler(ctx *gin.Context) {
	if c.started.Load() {
		ctx.JSON(http.StatusOK, gin.H{"status": "started"})
		return
	}
	ctx.JSON(http.StatusServiceUnavailable, gin.H{"status": "starting"})
}

File: src/shared/python/mall_common/health.py
"""Health check router for FastAPI services."""

from fastapi import APIRouter

router = APIRouter(prefix="/health", tags=["health"])

_ready = False
_started = False


def set_ready(ready: bool) -> None:
    global _ready
    _ready = ready


def set_started(started: bool) -> None:
    global _started
    _started = started


@router.get("/ready")
async def readiness():
    if _ready:
        return {"status": "ready"}
    from fastapi.responses import JSONResponse
    return JSONResponse(status_code=503, content={"status": "not_ready"})


@router.get("/live")
async def liveness():
    return {"status": "alive"}


@router.get("/startup")
async def startup():
    if _started:
        return {"status": "started"}
    from fastapi.responses import JSONResponse
    return JSONResponse(status_code=503, content={"status": "starting"})
